runtime/src/blueoil.cpp

```diff
@@ -15,6 +17,130 @@
 
 
 namespace blueoil {
+
+int Tensor::shapeVolume() {
+    return this->shapeVolume(m_shape);
+}
+int Tensor::shapeVolume(std::vector<int> shape) {
+    return std::accumulate(shape.begin(), shape.end(),
+			   1, std::multiplies<int>());
+}
+
+Tensor::Tensor(std::vector<int> shape)
+    : m_shape(std::move(shape)),
+      m_data(std::vector<float>(this->shapeVolume(shape), 0)) {
+}
+Tensor::Tensor(std::vector<int> shape, std::vector<float> data)
+    : m_shape(std::move(shape)),
+      m_data(std::move(data)) {
+}
+Tensor::Tensor(std::vector<int> shape, float *arr)
+    : m_shape(std::move(shape)),
+      m_data(std::vector<float>(arr, arr + this->shapeVolume(shape))) {
+}
+Tensor::Tensor(const Tensor &tensor)
+    : m_shape(tensor.m_shape),
+      m_data(tensor.m_data) {
+}
+std::vector<int> Tensor::shape() {
+    return m_shape;
+}
+std::vector<int> Tensor::shape() const {
+    return m_shape;
+}
+float* Tensor::data() {
+    return &(m_data[0]);
+}
+
+static void Tensor_shape_dump(std::vector<int> shape) {
+    std::cout << "shape:";
+    for (auto itr = shape.begin(); itr != shape.end(); ++itr) {
+        std::cout << *itr << " ";
+    }
+    std::cout << std::endl;
+}
+static void Tensor_data_dump(float *data, std::vector<int> shape) {
```

Comment:
Why do you always pass shape by value? :thinking:

Reply:
It recursively drops one dimension and calls Tensor_data_dump for dump arbitrary dimension data.
argument "shape" dimension diminishes as the recursion level of the call gets deeper.

Reply:
ok